### PR TITLE
Add knowledge distillation model and loss function support

### PR DESCRIPTION
**Pull request by mrT23**

### CODE REVIEW

**Comment by mrT23** [comment]
## PR Code Suggestions ✨


Latest suggestions up to 9c67326
<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Add temperature-scaled KD with correct KL usage</summary>

___

**Implement temperature scaling for knowledge distillation by dividing logits by a <br>temperature <code>T</code>. Also, scale the KL divergence loss by <code>T*T</code> to maintain gradient <br>magnitudes and refactor the duplicated loss calculation.**

[timm/utils/model_kd.py [62-74]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR62-R74)

```diff
-prob_s = torch.nn.functional.log_softmax(output, dim=-1)
+T_kd = getattr(args, "kd_temperature", 1.0)
+# student probability calculation with temperature
+prob_s = torch.nn.functional.log_softmax(output / T_kd, dim=-1)
 
 # teacher probability calculation
 with torch.no_grad():
     input_kd = model_kd.normalize_input(input, model)
     out_t = model_kd.model(input_kd.detach())
-    prob_t = torch.nn.functional.softmax(out_t, dim=-1)
+    prob_t = torch.nn.functional.softmax(out_t / T_kd, dim=-1)
 
-# adding KL loss
+# KL loss with temperature scaling; log_target must be False for soft targets
+kd_loss = torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean', log_target=False) * (T_kd * T_kd)
+
 if not args.use_kd_only_loss:
-    _loss += args.alpha_kd * torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean')
-else:  # only kd
-    _loss = args.alpha_kd * torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean')
+    _loss = _loss + args.alpha_kd * kd_loss
+else:
+    _loss = args.alpha_kd * kd_loss
```


- [ ] **Apply / Chat** 


<details><summary>Suggestion importance[1-10]: 9</summary>

__

Why: The suggestion introduces temperature scaling, a standard and crucial technique for effective knowledge distillation, and correctly implements the required loss scaling (`T*T`), which is a common omission. This significantly improves the correctness and performance of the KD implementation.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Avoid side effects and missing cfg errors</summary>

___

**In <code>normalize_input</code>, clone the <code>input</code> tensor before applying normalization <br>transforms to prevent modifying the original tensor in-place. Also, access the <br>student model's <code>default_cfg</code> safely to avoid errors if it's missing.**

[timm/utils/model_kd.py [36-57]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR36-R57)

```diff
 def normalize_input(self, input, student_model):
-    if hasattr(student_model, 'module'):
-        model_s = student_model.module
-    else:
-        model_s = student_model
+    model_s = student_model.module if hasattr(student_model, 'module') else student_model
 
-    mean_student = model_s.default_cfg['mean']
-    std_student = model_s.default_cfg['std']
+    cfg_s = getattr(model_s, 'default_cfg', {})
+    mean_student = cfg_s.get('mean', self.mean_model_kd)
+    std_student = cfg_s.get('std', self.std_model_kd)
 
     input_kd = input
     if mean_student != self.mean_model_kd or std_student != self.std_model_kd:
-        std = (self.std_model_kd[0] / std_student[0], self.std_model_kd[1] / std_student[1],
+        # work on a clone to avoid side effects on the original input
+        input_kd = input.clone()
+        std = (self.std_model_kd[0] / std_student[0],
+               self.std_model_kd[1] / std_student[1],
                self.std_model_kd[2] / std_student[2])
-        transform_std = T.Normalize(mean=(0, 0, 0), std=std)
-
-        mean = (self.mean_model_kd[0] - mean_student[0], self.mean_model_kd[1] - mean_student[1],
+        mean = (self.mean_model_kd[0] - mean_student[0],
+                self.mean_model_kd[1] - mean_student[1],
                 self.mean_model_kd[2] - mean_student[2])
-        transform_mean = T.Normalize(mean=mean, std=(1, 1, 1))
-
-        input_kd = transform_mean(transform_std(input))
+        transform_std = T.Normalize(mean=(0.0, 0.0, 0.0), std=std)
+        transform_mean = T.Normalize(mean=mean, std=(1.0, 1.0, 1.0))
+        input_kd = transform_mean(transform_std(input_kd))
 
     return input_kd
```


- [ ] **Apply / Chat** 


<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: The suggestion correctly identifies a potential side effect where the input tensor is modified in-place by the normalization transforms, which could affect other parts of the training loop. Cloning the input before normalization is a valid fix. It also improves robustness by safely accessing the model configuration.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Compile on target device and read cfg safely</summary>

___

**Move the teacher model to the target device (<code>cuda</code> or <code>cpu</code>) before applying <br><code>torch.compile</code> to prevent compilation issues. Additionally, access the <br><code>default_cfg</code> attribute safely to handle cases where the model is wrapped.**

[timm/utils/model_kd.py [23-33]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR23-R33)

```diff
-# compile model
-model_kd.cpu().eval()
+# place on target device first, then eval, then try compile
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model_kd = model_kd.to(device).eval()
 try:
     model_kd = torch.compile(model_kd)
     _logger.info(f"torch.compile applied successfully to KD model")
 except Exception as e:
     _logger.warning(f"torch.compile failed with error {e}, continuing KD model without torch compilation")
 
-self.model = model_kd.cuda()
-self.mean_model_kd = model_kd.default_cfg['mean']
-self.std_model_kd = model_kd.default_cfg['std']
+self.model = model_kd
+# access default_cfg from the (possibly wrapped) module safely
+cfg = getattr(model_kd, 'default_cfg', None)
+if cfg is None and hasattr(model_kd, 'module'):
+    cfg = getattr(model_kd.module, 'default_cfg', {})
+self.mean_model_kd = cfg.get('mean', (0.485, 0.456, 0.406))
+self.std_model_kd = cfg.get('std', (0.229, 0.224, 0.225))
```


- [ ] **Apply / Chat** 


<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: The suggestion correctly points out that a model should be moved to its target device before `torch.compile` is called, preventing potential issues. It also adds robust handling for accessing `default_cfg` on a possibly wrapped model, which prevents potential runtime errors.


</details></details></td><td align=center>Medium

</td></tr>
<tr><td align="center" colspan="2">

- [ ] More 

</td><td></td></tr></tbody></table>

___

#### Previous suggestions
<details><summary>Suggestions up to commit 8bee03c</summary>
<br><table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=2>Possible issue</td>
<td>



<details><summary>Fix missing import and correct normalization</summary>

___

**Import <code>torchvision.transforms as T</code> to resolve a <code>NameError</code>. Additionally, correct <br>the input normalization logic by using a single, mathematically sound <br><code>T.Normalize</code> operation to prevent providing incorrect data to the teacher model.**

[timm/utils/model_kd.py [35-56]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR35-R56)

```diff
+import torchvision.transforms as T
+...
 def normalize_input(self, input, student_model):
     if hasattr(student_model, 'module'):
         model_s = student_model.module
     else:
         model_s = student_model
 
     mean_student = model_s.default_cfg['mean']
     std_student = model_s.default_cfg['std']
 
+    # By default, assume input is currently normalized for the student.
     input_kd = input
     if mean_student != self.mean_model_kd or std_student != self.std_model_kd:
-        std = (self.std_model_kd[0] / std_student[0], self.std_model_kd[1] / std_student[1],
-               self.std_model_kd[2] / std_student[2])
-        transform_std = T.Normalize(mean=(0, 0, 0), std=std)
-
-        mean = (self.mean_model_kd[0] - mean_student[0], self.mean_model_kd[1] - mean_student[1],
-                self.mean_model_kd[2] - mean_student[2])
-        transform_mean = T.Normalize(mean=mean, std=(1, 1, 1))
-
-        input_kd = transform_mean(transform_std(input))
+        # Transform from student norm to teacher norm:
+        # x_t = (x_s * std_s + mean_s - mean_t) / std_t
+        # Which equals: x_s * (std_s / std_t) + (mean_s - mean_t) / std_t
+        scale = (
+            std_student[0] / self.std_model_kd[0],
+            std_student[1] / self.std_model_kd[1],
+            std_student[2] / self.std_model_kd[2],
+        )
+        shift = (
+            (mean_student[0] - self.mean_model_kd[0]) / self.std_model_kd[0],
+            (mean_student[1] - self.mean_model_kd[1]) / self.std_model_kd[1],
+            (mean_student[2] - self.mean_model_kd[2]) / self.std_model_kd[2],
+        )
+        # T.Normalize computes (x - mean) / std; to achieve scale*x + shift, set:
+        # mean = -shift/scale, std = 1/scale
+        # Guard against division by zero
+        eps = 1e-8
+        std = tuple(max(s, eps) for s in scale)
+        mean = tuple(-shift[i] / (std[i]) for i in range(3))
+        transform = T.Normalize(mean=mean, std=tuple(1.0 / s for s in std))
+        input_kd = transform(input)
 
     return input_kd
```






<details><summary>Suggestion importance[1-10]: 10</summary>

__

Why: This suggestion fixes a fatal `NameError` due to a missing import of `T` and corrects a silent but critical bug in the normalization logic, which would have provided incorrect data to the teacher model.


</details></details></td><td align=center>High

</td></tr><tr><td>



<details><summary>Fix device handling and compilation order</summary>

___

**Move the teacher model to the target device before <code>torch.compile</code> and avoid <br>hardcoding <code>.cuda()</code> to support CPU-only execution and prevent performance loss <br>from recompilation.**

[timm/utils/model_kd.py [8-32]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR8-R32)

```diff
 class build_kd_model(nn.Module):
-    def __init__(self, args):
+    def __init__(self, args, device=None):
         super(build_kd_model, self).__init__()
 
         _logger.info(f"Creating KD model: from '{args.kd_model_name}'")
         in_chans = 3
         if args.in_chans is not None:
             in_chans = args.in_chans
         model_kd = create_model(
             model_name=args.kd_model_name,
             num_classes=args.num_classes,
             pretrained=True,
-            in_chans=in_chans)
+            in_chans=in_chans
+        )
 
-        # compile model
-        model_kd.cpu().eval()
+        # choose device: follow provided device or infer from CUDA availability
+        device = device if device is not None else (torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu'))
+        model_kd = model_kd.to(device).eval()
+
+        # compile model (only if available and on CUDA to avoid CPU graph fallbacks)
         try:
-            model_kd = torch.compile(model_kd)
-            _logger.info(f"torch.compile applied successfully to KD model")
+            if hasattr(torch, "compile"):
+                model_kd = torch.compile(model_kd)
+                _logger.info("torch.compile applied successfully to KD model")
+            else:
+                _logger.info("torch.compile not available, continuing without compilation for KD model")
         except Exception as e:
             _logger.warning(f"torch.compile failed with error {e}, continuing KD model without torch compilation")
 
-        self.model = model_kd.cuda()
+        self.model = model_kd
         self.mean_model_kd = model_kd.default_cfg['mean']
         self.std_model_kd = model_kd.default_cfg['std']
```


 


<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: The suggestion correctly identifies that hardcoding `.cuda()` is a bug for CPU-only environments and that compiling before moving the model to the correct device is inefficient, causing recompilation.


</details></details></td><td align=center>Medium

</td></tr><tr><td rowspan=1>General</td>
<td>



<details><summary>Add temperature and device safeguards</summary>

___

**Implement temperature scaling for both student and teacher logits in the <br>KL-divergence loss calculation to improve training stability, a standard <br>practice for knowledge distillation.**

[timm/utils/model_kd.py [59-75]](https://github.com/mrT23/pytorch-image-models/pull/5/files#diff-419e9f11516731c46fd157799db4c0829a23ce2f6bbd2041ea5322686f95430fR59-R75)

```diff
 def add_kd_loss(_loss, output, input, model, model_kd, args):
-    # student probability calculation
-    prob_s = torch.nn.functional.log_softmax(output, dim=-1)
+    T_kd = getattr(args, "kd_temperature", 1.0)
+
+    # student log-probabilities with temperature
+    prob_s = torch.nn.functional.log_softmax(output / T_kd, dim=-1)
 
     # teacher probability calculation
     with torch.no_grad():
         input_kd = model_kd.normalize_input(input, model)
+        # ensure same device
+        input_kd = input_kd.to(next(model_kd.model.parameters()).device)
         out_t = model_kd.model(input_kd.detach())
-        prob_t = torch.nn.functional.softmax(out_t, dim=-1)
+        prob_t = torch.nn.functional.softmax(out_t / T_kd, dim=-1)
 
-    # adding KL loss
+    # guard against class mismatch
+    if prob_s.shape[-1] != prob_t.shape[-1]:
+        raise RuntimeError(f"KD class mismatch: student={prob_s.shape[-1]} teacher={prob_t.shape[-1]}")
+
+    kd = torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean') * (T_kd * T_kd)
     if not args.use_kd_only_loss:
-        _loss += args.alpha_kd * torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean')
-    else:  # only kd
-        _loss = args.alpha_kd * torch.nn.functional.kl_div(prob_s, prob_t, reduction='batchmean')
+        _loss = _loss + args.alpha_kd * kd
+    else:
+        _loss = args.alpha_kd * kd
 
     return _loss
```


 


<details><summary>Suggestion importance[1-10]: 7</summary>

__

Why: The suggestion correctly points out the lack of temperature scaling, a standard practice in knowledge distillation for better training stability, and proposes a valid implementation.


</details></details></td><td align=center>Medium

</td></tr>
<tr><td align="center" colspan="2">

 

</td><td></td></tr></tbody></table>

</details>

**Comment by mrT23** [comment]
**[Persistent suggestions](https://github.com/mrT23/pytorch-image-models/pull/5#issuecomment-3405930341)** updated to latest commit 9c67326
